cadquery/occ_impl/solver.py

```diff
@@ -1,17 +1,99 @@
-from typing import Tuple, Union, Any, Callable, List, Optional, Dict
+from typing import (
+    List,
+    Tuple,
+    Union,
+    Any,
+    Callable,
+    Optional,
+    Dict,
+    Literal,
+    cast as tcast,
+    Type,
+)
 from nptyping import NDArray as Array
-
-from numpy import array, eye, zeros, pi
+from math import radians
+from typish import instance_of, get_type
+from numpy import array, eye, pi
 import nlopt
 
-from OCP.gp import gp_Vec, gp_Pln, gp_Dir, gp_Pnt, gp_Trsf, gp_Quaternion
+from OCP.gp import (
+    gp_Vec,
+    gp_Pln,
+    gp_Dir,
+    gp_Pnt,
+    gp_Trsf,
+    gp_Quaternion,
+    gp_XYZ,
+    gp_Lin,
+    gp_Intrinsic_XYZ,
+)
+from OCP.BRepTools import BRepTools
+from OCP.Precision import Precision
+
+from .geom import Location, Vector, Plane
+from .shapes import Shape, Face, Edge, Wire
+from ..types import Real
 
-from .geom import Location
+# type definitions
+
+NoneType = type(None)
 
 DOF6 = Tuple[float, float, float, float, float, float]
-ConstraintMarker = Union[gp_Pln, gp_Dir, gp_Pnt]
+ConstraintMarker = Union[gp_Pln, gp_Dir, gp_Pnt, gp_Lin, None]
+
+UnaryConstraintKind = Literal[
+    "Fixed", "FixedPoint", "FixedAxis", "FixedRotation", "FixedRotationAxis"
+]
+BinaryConstraintKind = Literal["Plane", "Point", "Axis", "PointInPlane", "PointOnLine"]
+ConstraintKind = Literal[
+    "Plane",
+    "Point",
+    "Axis",
+    "PointInPlane",
+    "Fixed",
+    "FixedPoint",
+    "FixedAxis",
+    "PointOnLine",
+    "FixedRotation",
+    "FixedRotationAxis",
+]
+
+# (arity, marker types, param type, conversion func)
+ConstraintInvariants = {
+    "Point": (2, (gp_Pnt, gp_Pnt), Real, None),
+    "Axis": (
+        2,
+        (gp_Dir, gp_Dir),
+        Real,
+        lambda x: radians(x) if x is not None else None,
+    ),
+    "PointInPlane": (2, (gp_Pnt, gp_Pln), Real, None),
+    "PointOnLine": (2, (gp_Pnt, gp_Lin), Real, None),
+    "Fixed": (1, (None,), Type[None], None),
+    "FixedPoint": (1, (gp_Pnt,), Tuple[Real, Real, Real], None),
+    "FixedAxis": (1, (gp_Dir,), Tuple[Real, Real, Real], None),
+    "FixedRotationAxis": (
+        1,
+        (None,),
+        Tuple[int, Real],
+        lambda x: (x[0], radians(x[1])),
+    ),
+}
+
+# translation table for compound constraints {name : (name, ...), converter}
+CompoundConstraints: Dict[
+    ConstraintKind, Tuple[Tuple[ConstraintKind, ...], Callable[[Any], Tuple[Any, ...]]]
+] = {
+    "Plane": (("Axis", "Point"), lambda x: (x, 0)),
```

Comment:
```suggestion
    "Plane": (("Axis", "Point"), lambda x: (radians(x) if x is not None else None, 0)),
```
This is to apply the conversion func to Axis param when part of the Plane compound constraint.  Will look more tomorrow!